### PR TITLE
Feat/anti diagonal gates

### PR DESCRIPTION
**Pull request by fSosaRey**
Merge from branch feat/abstract-phase-gate, and implemented changes mentioned in https://github.com/anyonlabs/Snowflake.jl/pull/73#discussion_r1143438226

### CODE REVIEW

**Comment by fSosaRey** [comment]
coverage drops and documentation fails because Circuits cannot handle Gates and AbstractGates at once. will recover once all AbstractGates are implemented

File: test/antiDiagonalGate.jl [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        ])
+
+    @test anti_diag_op*anti_diag_op≈ result
+    @test Operator(anti_diag_op)*anti_diag_op≈ result
+    @test anti_diag_op*Operator(anti_diag_op)≈ result
+    
+    # Exponentiation
+    θ=π 
+    anti_diag_op=AntiDiagonalOperator([1,1])
+    @test (exp(-im*θ/2*anti_diag_op)).data ≈ [[0.,-im] [-im,0.]]
```
Comment (plerouxanyon) [inline test/antiDiagonalGate.jl:47]: Inline theta or define the result in terms of theta.

File: test/antiDiagonalGate.jl [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        0.0 - 8.0im,
+        0.0 + 1.0im,
+        0.0 + 2.0im,
+        0.0 + 3.0im,
+        0.0 + 4.0im
+    ])
+
+    @test ψ ≈ ψ_result
+
+    @test adjoint(y_operator).data==ComplexF64[im,-im]
```
Comment (plerouxanyon) [inline test/antiDiagonalGate.jl:84]: Can we test this without reaching into the implementation details? (Test using only the public interface)

File: test/diagonalGates.jl [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    
+    ψ_z = Ket([v for v in 1:2^qubit_count])
+
+    ZGate=sigma_z(target)
+    apply_gate!(ψ_z, ZGate)
+
+    @test ψ≈ψ_z
+
+    # Ctor from LinearAlgebra.Adjoint(DiagonalOperator{N,T})
+    @test adjoint(phase_gate_operator).data≈get_operator(Snowflake.phase_shift(target,-ϕ)).data
```
Comment (plerouxanyon) [inline test/diagonalGates.jl:51]: Again, can we test this without looking at the implementation detail?

File: test/diagonalGates.jl [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    @test_throws DomainError apply_gate!(ψ, T_gate_error)
+
+end
+
+@testset "Diagonal Gate: N targets" begin
+
+
+    ######################################################
+
+    diagonalGate_N_targets(target_list::Vector{<:Integer}) =
```
Comment (plerouxanyon) [inline test/diagonalGates.jl:158]: Why is this necessary? It looks like we are testing a test.

File: test/diagonalOperator.jl [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    diag_op=DiagonalOperator([1.,2.])
+
+    # Base.:* specialization
+
+    @test (Operator(diag_op)*diag_op)≈ DiagonalOperator([v^2 for v in Vector(diag_op.data)])
+    @test (diag_op*Operator(diag_op))≈ DiagonalOperator([v^2 for v in Vector(diag_op.data)])
+
+    # Exponentiation
+    θ=π
+    @test (exp(-im*θ/2*diag_op)).data ≈ [-im,-1.]
```
Comment (plerouxanyon) [inline test/diagonalOperator.jl:21]: Again, inline theta or define result in terms of theta

File: test/testFunctions.jl [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+function test_inverse(gate::Snowflake.AbstractGate)
+    inverse_gate=get_inverse(gate)
+    target_count=length(Snowflake.get_connected_qubits(gate))
+
+    return( get_operator(gate)*get_operator(inverse_gate) ≈ eye(target_count) )
+end
+
+function make_array(
+    dim::Integer,
```
Comment (plerouxanyon) [inline test/testFunctions.jl:18]: This function seems unnecessary. Can we inline this function?

File: src/core/qobj.jl [hunk 7]
```diff
@@ -109,22 +109,140 @@ Underlying data Matrix{ComplexF64}:
 Alternatively:
 ```jldoctest
 julia> z = Snowflake.sigma_z()  #sigma_z is a defined function in Snowflake
-(2, 2)-element Snowflake.Operator:
-Underlying data Matrix{ComplexF64}:
-1.0 + 0.0im    0.0 + 0.0im
-0.0 + 0.0im    -1.0 + 0.0im
+(2,2)-element Snowflake.DiagonalOperator:
+Underlying data type: ComplexF64:
+1.0 + 0.0im    .
```
Comment (plerouxanyon) [inline src/core/qobj.jl:114]: Can we align the printout of DiagonalOperator?

File: src/core/qobj.jl [hunk 8]
```diff
@@ -870,52 +1094,69 @@ end
...
-end
+commute(A::Operator, B::Operator) = A*B-B*A
+
+# generic cases
+commute(A::AbstractOperator, B::Operator)= commute(Operator(A),B)
+commute(A::Operator, B::AbstractOperator)= commute(A,Operator(B))
+commute(A::AbstractOperator, B::AbstractOperator)= commute(Operator(A),Operator(B))
+
+# specializations
+commute(A::DiagonalOperator, B::DiagonalOperator)= A*B-B*A
```
Comment (plerouxanyon) [inline src/core/qobj.jl:1126]: How is this different from the base commute?

File: src/core/qobj.jl [hunk 9]
```diff
@@ -870,52 +1094,69 @@ end
...
-end
+anticommute(A::Operator, B::Operator)= A*B+B*A
+
+# generic cases
+anticommute(A::AbstractOperator, B::Operator)=anticommute(Operator(A),B)
+anticommute(A::Operator, B::AbstractOperator)=anticommute(A,Operator(B))
+anticommute(A::AbstractOperator, B::AbstractOperator)=anticommute(Operator(A),Operator(B))
+
+# specializations
+anticommute(A::DiagonalOperator, B::DiagonalOperator)=A*B+B*A
```
Comment (plerouxanyon) [inline src/core/qobj.jl:1158]: Again, how is this different from the base anticommute?

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: src/core/qobj.jl [hunk 10]
```diff
@@ -507,9 +649,12 @@ function get_embed_operator(op::Operator, target_body_index::Int, system::MultiB
     return result
 end
 
-get_embed_operator(op::DiagonalOperator, target_body_index::Int, system::MultiBodySystem)=
+get_embed_operator(op::AbstractOperator, target_body_index::Int, system::MultiBodySystem)=
     get_embed_operator(Operator(op), target_body_index, system)
 
+get_matrix(op::AbstractOperator) =
```
Comment (plerouxanyon) [inline src/core/qobj.jl:655]: Could you add a test for this?

File: src/core/qobj.jl [hunk 11]
```diff
@@ -138,18 +148,33 @@ A structure representing a diagonal quantum operator (i.e. a complex matrix, wit
 # Examples
 ```jldoctest
 julia> z = Snowflake.DiagonalOperator([1.0,-1.0])
+<<<<<<< HEAD
```
Comment (plerouxanyon) [inline src/core/qobj.jl:151]: This looks like a failed merge or rebase.

File: src/core/quantum_gate.jl [hunk 12]
```diff
@@ -417,7 +463,10 @@ T = \\begin{bmatrix}
     \\end{bmatrix}.
 ```
 """
-pi_8(T::Type{<:Complex}=ComplexF64) = DiagonalOperator{2,T}(T[1.,exp(im*pi/4.0)])
+pi_8(T::Type{<:Complex}=ComplexF64) = DiagonalOperator(T[1.,exp(im*pi/4.0)])
+
+pi_8_diag(T::Type{<:Complex}=ComplexF64) = DiagonalOperator{2,T}(T[1.,exp(im*pi/4.0)])
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:468]: This operator should be removed.

**Comment by plerouxanyon** [comment]
The documentation is also failing.

**Comment by plerouxanyon** [comment]
Just mentioning here for information: The docs and coverage are failing due to the missing tests and failing documentation of the circuit since the circuit cannot use abstract gates yet.

**Pull request merged**